File: app/controllers/Coupon.scala
package controllers

import controllers.common.BaseController
import com.google.inject.Inject
import utility.Authorized
import service.common.sql.CouponService
import play.api.libs.json.Json
import model.Coupon.writes

class Coupon @Inject()(couponService: CouponService) extends BaseController {
  def list(userId: Int) = Authorized(userIds = Seq(userId), roles = Seq("admin")) {
    request =>
      val list = couponService.list(userId)
      Ok(Json.toJson(list))
  }

  def get(number: String, userId: Int) = Authorized(userIds = Seq(userId), roles = Seq("admin")) {
    request =>
      couponService.get(number, userId) match {
        case Some(coupon) => Ok(Json.toJson(coupon))
        case _ => BadRequest("Could not find the coupon")
      }
  }
}

File: public/app/scripts/services/http/Coupon.js
(function (app, routes) {
  CouponService.$inject = ['$http'];
  function CouponService($http) {
    this.$http = $http;
  }

  CouponService.prototype = {
    list: function(userId) {
      return this.$http(routes.Coupon.list(userId)).then(function(response){
        return response.data;
      });
    },
    get: function(number, userId) {
      return this.$http(routes.Coupon.get(number, userId)).then(function(response){
        return response.data;
      });
    }
  };
  app.service('CouponService', CouponService);
  return CouponService;
}(angular.module('starwallet'), jsRoutes.controllers));